src/NServiceBus.AmazonSQS/Batcher.cs

```diff
@@ -0,0 +1,57 @@
+namespace NServiceBus.Transports.SQS
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using AmazonSQS;
+
+    static class Batcher
+    {
+        public static IReadOnlyList<BatchEntry> Batch(IEnumerable<PreparedMessage> preparedMessages)
+        {
+            var allBatches = new List<BatchEntry>();
+            var currentDestinationBatches = new Dictionary<string, PreparedMessage>();
+
+            var groupByDestination = preparedMessages.GroupBy(m => m.QueueUrl, StringComparer.OrdinalIgnoreCase);
+            foreach (var group in groupByDestination)
+            {
+                PreparedMessage firstMessage = null;
+                var payloadSize = 0;
+                for (var i = 0; i < group.Count(); i++)
```

Comment:
nitpicking/curiosity: given that `Count()` enumerates all items to retrieve the count, and then a for is performed, isn't a `foreach` a better alternative?

Reply:
Yeah will switch